src/tracedAtomic.ml

```diff
@@ -208,6 +224,7 @@ let do_run init_func init_schedule =
                         |> Seq.filter (fun (_,proc) -> not proc.finished)
                         |> Seq.map (fun (id,_) -> id)
                         |> IntSet.of_seq in
+  CCVector.iter (fun proc -> proc.discontinue_f ()) processes;
```

Comment:
This seems like a fairly straightforward fix to me (i.e. discontinue the continuations after being done with them).

Reply:
Hmm I think this is a solution to the wrong problem. The reason why we end up with pending continuations is that the evaluation of a program `A;B;C;D` by dscheck proceeds as:

```
A  abort, restart from scratch
A;B  abort, restart
A;B;C  abort, restart
A;B;C;D  finally ok, will attempt to backtrack next
```

while it should just do `A;B;C;D` from the get-go (... but I'm totally biased because that was the first optimization of https://github.com/ocaml-multicore/dscheck/pull/3 so I'm not too keen on this consolidation of the abort-and-retry strategy :/ )

Reply:
I disagree.

* Fixing the memory leak gets us 90-something percent there. It reduces memory complexity from exponential to low polynomial and addresses the pressing issue at hand, i.e. ooms on non-trivial tests or with multiple trivial tests at once. That's with marginal extra code to maintain and imposing no restrictions on core algorithm. The optimization you describe is going to yield a linear gain in time on algorithms with exponential time complexity. 

* If anything, this fix provides a baseline for evaluation of future optimizations. Without it, any improvement will look great but really you're just measuring the fix of a memory leak.

Reply:
I guess it's a matter of point of view, but I find the code more complex if it's covering up the leak rather than not having it in the first place. It adds code that shouldn't be necessary. There are multiple ways of fixing a bug, so it's worth reflecting on which direction this is pushing the codebase towards.

I don't know how to make `dscheck` fast enough to be usable while keeping this quadratic evaluation model that is the issue in the first place. I get 0.8s on `test_list.exe` with your PR, and 0.4s with only the first patch of #3 that avoids the abort-and-restart leak (and 0.01s with all #3, but I understand that the late optims are harder to validate as being correct). I'm happy with #3 being closed as I don't have time, but it's a bit sad if the next step is to revert this PR to make progress on dscheck performances.

Reply:
> There are multiple ways of fixing a bug, so it's worth reflecting on which direction this is pushing the codebase towards.

This is very much where I'm coming from. Discontinuing continuations is the smallest step we can take towards correct OCaml code. I don't think that reverting in either of the patches is going to be a big job, but it's going to be far easier to revert fix of a memory leak, than backpedal on locking core algorithm into DFS and sacrificing pure `do_run`. 

Reduction to 0.4s is admirable but dscheck is still going to explode on n+2 on a complex test rather than n. It's not going to give us a qualitative difference. Conversely, if we modernize pruning in core algo, we might not even need per-state optimizations. 

We can still look at the first commit in #3 because it's arguably the best we have. But we should not hold off on stabilizing dscheck testsuites.

Reply:
I mean, to be clear, you don't need my approval to move forward with your plan, I'm just a passer-by chipping into the review process. I understand that you only care about the memory leak since it's a pressing issue. I admit I'm stubborn, my opinion is still that taking a detour to treat the symptom isn't the fix I would recommend here.

Regarding the exponential complexity, we just need dscheck to be able to run our tests (which are bounded by how slow dscheck is). I look at its complexity as *"dscheck must validate `N` traces each composed of roughly `S` steps"*, so `O(N*S)` vs `O(N*S²)` does have an impact in practice. Sure `N` hides an exponential and implies that we'll never be able to run arbitrarily complex tests. But even with a small `S=30`, we are going to use dscheck differently if it responds in 1min or 30min later.

If you have some time, I would appreciate if you could implement this optimization for comparison. If it turns out that you like it less than this PR, then so be it!

Reply:
Hmm... 🤔  I'm not sure that I understand the objection.  Does the fix prevent future improvements?

Maybe I need to take a closer look at the code and different change proposals.

Reply:
Sounds good, I'm going to leave it open.

>  I understand that you only care about the memory leak since it's a pressing issue.

Perhaps I have not communicated clearly, but it's really not the case. I do believe that fixing this memory leak is a step in the right direction, and it's not significantly hindering future improvements. 

> Regarding the exponential complexity, we just need dscheck to be able to run our tests (which are bounded by how slow dscheck is). I look at its complexity as "dscheck must validate N traces each composed of roughly S steps", so O(N*S) vs O(N*S²) does have an impact in practice. Sure N hides an exponential and implies that we'll never be able to run arbitrarily complex tests. But even with a small S=30, we are going to use dscheck differently if it responds in 1min or 30min later.

N being exponential here has a stronger implication: N-2 is going to be trivial in any case, while N+2 is going to explode in any case. 

With that, if we were going to get a 97% time reduction on all test I would be much more optimistic to jump to first commit in #3 but I do not think we're going to get such an improvement in the general case. This patch optimizes processing of non-leaf nodes. But it still requires at least one execution per leaf node. Leaf nodes may very well be majority of the nodes in our graph. Especially if there's more than 2 threads or tested algorithm's accesses are not largely on disjoint locations.  

Like, I'm not even opposed to #3. We can always add it afterwards. In particular, if we want to sacrifice some flexibility in the core algo for extra speed. I don't think the two solutions are necessarily in conflict.

Reply:
> I do believe that fixing this memory leak is a step in the right direction

Ha yeah, I see where we disagree, my objection is that this fix will need to be reverted to make progress (so I would prefer if the origin of the leak was addressed instead)

Reply:
This PR addresses precisely the origin of the leak. Continuations are cancellable for a reason. 

Yes, we can force the algorithm to execute continuations to the end instead but that's not in any way mutually exclusive with discontinuing ones we don't need.

Reply:
Also doesn't the same argument apply in reverse? #3 is a far more restricting change to the core algo and we might end up needing to revert it as soon as we try to improve on the exponential.

Reply:
I mean, your PR is cleaning up after the leak. The origin of the leak is that the continuations are not executed fully. There's nothing to discontinue once this is addressed. Cancelling the continuations is incompatible with running them to the end. I don't know that there's a core algorithm to preserve, I only know that we are wasting resources by aborting the continuations early.

But I'm probably not going to convince you at this point :P It seems more productive if you carry on with your plan and revisit the "run continuations to the end" later.

Reply:
> Cancelling the continuations is incompatible with running them to the end. 

This PR terminates continuations in both ways. Likewise, if #3 needed to sometimes step off depth-first (e.g. for particular op), it would have to start discontinuing at those points only.

Reply:
> I mean, your PR is cleaning up after the leak. The origin of the leak is that the continuations are not executed fully.

I disagree. A memory leak is precisely the lack of clean-up of resources that need to be cleaned up. Likewise `free` prevents the leaks, rather than clean up after them.

Reply:
Well the search order got nothing to do with wasting runs by cancelling them. Issue https://github.com/ocaml-multicore/dscheck/issues/9 will be a valid reason for cleaning up though!

Anyway, I'm sorry if I was too blunt in my feedback. We can keep arguing but I don't think we understand each other. My worst outcome has already been reached if you've convinced yourself that dscheck performances isn't worth improving.. Solving sudoku is exponential but there are tricks to speed it up! I really hope you will go on to write some cool optimizations, the code will tell you what's slow and it's a lot of fun :)